File: services/clustering-service/app/schemas/escalation.py
"""Pydantic schemas for escalation API responses.

This module defines response schemas for the Intelligence Interpretation Layer
escalation endpoints. Includes domain-level escalation scores, combined metrics,
market regime data, correlation alerts, and cluster-level escalation details.
"""

from datetime import datetime
from decimal import Decimal
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_serializer


class DomainEscalationResponse(BaseModel):
    """Escalation data for a single domain.

    Represents the escalation level and score for one of three domains:
    geopolitical, military, or economic.

    Attributes:
        domain: Domain name (geopolitical, military, or economic)
        level: Escalation level 1-5 (1=low, 5=critical)
        score: Normalized score 0.000-1.000
        confidence: Confidence in the assessment (0.0-1.0)
    """

    model_config = ConfigDict(
        from_attributes=True,
    )

    domain: str = Field(..., description="Domain: geopolitical, military, or economic")
    level: int = Field(..., ge=1, le=5, description="Escalation level 1-5")
    score: Decimal = Field(..., ge=0, le=1, description="Normalized score 0.000-1.000")
    confidence: float = Field(..., ge=0, le=1, description="Confidence in assessment")

    @field_serializer("score")
    def serialize_score(self, value: Decimal) -> str:
        """Serialize Decimal as string for JSON output."""
        return str(value)


class CorrelationAlertResponse(BaseModel):
    """Active correlation alert.

    Represents a detected correlation between news escalation and market regime.

    Attributes:
        id: Unique alert identifier
        correlation_type: CONFIRMATION, DIVERGENCE, or EARLY_WARNING
        fmp_regime: Current FMP market regime (RISK_ON, RISK_OFF, TRANSITIONAL)
        escalation_level: News escalation level at time of detection (1-5)
        confidence: Confidence score for this correlation
        reasoning: Human-readable explanation of the correlation
        detected_at: When this correlation was detected
        expires_at: When this alert expires
        related_cluster_count: Number of clusters related to this alert
    """

    model_config = ConfigDict(
        from_attributes=True,
    )

    id: UUID
    correlation_type: str = Field(
        ..., description="CONFIRMATION, DIVERGENCE, or EARLY_WARNING"
    )
    fmp_regime: str = Field(..., description="RISK_ON, RISK_OFF, or TRANSITIONAL")
    escalation_level: int = Field(..., ge=1, le=5)
    confidence: Decimal
    reasoning: Optional[str] = None
    detected_at: datetime
    expires_at: Optional[datetime] = None
    related_cluster_count: int = 0

    @field_serializer("confidence")
    def serialize_confidence(self, value: Decimal) -> str:
        """Serialize Decimal as string for JSON output."""
        return str(value)


class RegimeStateResponse(BaseModel):
    """Current FMP market regime.

    Represents the current state of the market as reported by the FMP service.

    Attributes:
        regime: Current regime (RISK_ON, RISK_OFF, TRANSITIONAL)
        confidence: Confidence in the regime assessment
        vix_level: Current VIX level (if available)
        fear_greed_index: Fear & Greed index 0-100 (if available)
        timestamp: When the regime was last updated
    """

    model_config = ConfigDict(
        from_attributes=True,
    )

    regime: str = Field(..., description="RISK_ON, RISK_OFF, or TRANSITIONAL")
    confidence: float = Field(..., ge=0, le=1)
    vix_level: Optional[float] = None
    fear_greed_index: Optional[int] = Field(None, ge=0, le=100)
    timestamp: Optional[datetime] = None


class EscalationSummaryResponse(BaseModel):
    """Complete escalation summary for dashboard.

    Aggregates escalation data across all domains, combined metrics,
    market regime correlation, and active alerts.

    Attributes:
        geopolitical: Geopolitical domain escalation data
        military: Military domain escalation data
        economic: Economic domain escalation data
        combined_level: Maximum of domain levels (1-5)
        combined_score: Average of domain scores
        market_regime: Current FMP market regime (if available)
        correlation_alerts: List of active correlation alerts
        cluster_count: Number of clusters analyzed
        calculated_at: When this summary was calculated
    """

    model_config = ConfigDict(
        from_attributes=True,
    )

    # Domain escalation levels
    geopolitical: DomainEscalationResponse
    military: DomainEscalationResponse
    economic: DomainEscalationResponse

    # Combined metrics
    combined_level: int = Field(..., ge=1, le=5, description="Max of domain levels")
    combined_score: Decimal = Field(..., description="Average of domain scores")

    # Market correlation
    market_regime: Optional[RegimeStateResponse] = None
    correlation_alerts: List[CorrelationAlertResponse] = Field(default_factory=list)

    # Metadata
    cluster_count: int = Field(..., description="Number of clusters analyzed")
    calculated_at: datetime

    @field_serializer("combined_score")
    def serialize_combined_score(self, value: Decimal) -> str:
        """Serialize Decimal as string for JSON output."""
        return str(value)


class SignalDetailResponse(BaseModel):
    """Individual signal breakdown for escalation calculation.

    Represents a single signal from one of three sources (embedding, content,
    or keywords) used in the escalation calculation for a cluster.

    Attributes:
        source: Signal source identifier (embedding, content, or keywords)
        level: Escalation level 1-5 contributed by this signal
        confidence: Confidence in this signal's contribution (0.0-1.0)
        matched_anchor_id: UUID of the matched anchor (for embedding/keyword signals)
        matched_keywords: List of matched keywords (for keyword signals)
        reasoning: Human-readable explanation of the signal
    """

    model_config = ConfigDict(
        from_attributes=True,
    )

    source: str = Field(..., description="Signal source: embedding, content, or keywords")
    level: int = Field(..., ge=1, le=5, description="Escalation level 1-5")
    confidence: float = Field(..., ge=0, le=1, description="Signal confidence 0.0-1.0")
    matched_anchor_id: Optional[UUID] = Field(
        None, description="UUID of matched anchor point"
    )
    matched_keywords: Optional[List[str]] = Field(
        None, description="List of matched keywords"
    )
    reasoning: Optional[str] = Field(None, description="Explanation of signal")


class ClusterEscalationDetailResponse(BaseModel):
    """Detailed escalation data for a single cluster.

    Provides comprehensive escalation information for a specific cluster,
    including domain-level breakdowns and individual signal contributions.

    Attributes:
        cluster_id: UUID of the cluster
        cluster_title: Title/label of the cluster
        article_count: Number of articles in the cluster
        geopolitical: Geopolitical domain escalation
        military: Military domain escalation
        economic: Economic domain escalation
        combined_level: Maximum level across domains (1-5)
        combined_score: Weighted average score
        geopolitical_signals: Signal breakdown for geopolitical domain
        military_signals: Signal breakdown for military domain
        economic_signals: Signal breakdown for economic domain
        escalation_calculated_at: When escalation was last calculated
        created_at: Cluster creation timestamp
    """

    model_config = ConfigDict(
        from_attributes=True,
    )

    cluster_id: UUID
    cluster_title: str
    article_count: int

    # Domain breakdowns
    geopolitical: DomainEscalationResponse
    military: DomainEscalationResponse
    economic: DomainEscalationResponse

    # Combined metrics
    combined_level: int = Field(..., ge=1, le=5, description="Max of domain levels")
    combined_score: Decimal = Field(..., description="Average of domain scores")

    # Signal breakdowns per domain
    geopolitical_signals: List[SignalDetailResponse] = Field(default_factory=list)
    military_signals: List[SignalDetailResponse] = Field(default_factory=list)
    economic_signals: List[SignalDetailResponse] = Field(default_factory=list)

    # Metadata
    escalation_calculated_at: Optional[datetime] = None
    created_at: datetime

    @field_serializer("combined_score")
    def serialize_combined_score(self, value: Decimal) -> str:
        """Serialize Decimal as string for JSON output."""
        return str(value)

File: frontend/src/features/intelligence/types/escalation.ts
/**
 * Escalation Types for Intelligence Interpretation Layer
 *
 * TypeScript types matching backend Pydantic schemas in clustering-service.
 * Used for escalation scoring across three domains (geopolitical, military, economic)
 * with FMP market regime correlation.
 *
 * @module features/intelligence/types/escalation
 */

// =============================================================================
// Base Types
// =============================================================================

/**
 * Market regime type from FMP service
 */
export type RegimeType = 'RISK_ON' | 'RISK_OFF' | 'TRANSITIONAL' | 'UNKNOWN';

/**
 * Correlation alert type between market regime and news escalation
 */
export type AlertType = 'CONFIRMATION' | 'DIVERGENCE' | 'EARLY_WARNING';

/**
 * Escalation level (1-5 scale)
 * 1: Routine, 2: Elevated, 3: Significant, 4: High, 5: Critical
 */
export type EscalationLevel = 1 | 2 | 3 | 4 | 5;

/**
 * Domain for escalation scoring
 */
export type EscalationDomain = 'geopolitical' | 'military' | 'economic';

// =============================================================================
// Core Interfaces
// =============================================================================

/**
 * Current market regime status from FMP
 */
export interface RegimeStatus {
  /** Current regime type */
  type: RegimeType;
  /** Regime confidence/strength score (0-1) */
  score: number;
  /** ISO timestamp when this regime started */
  since?: string;
}

/**
 * Financial market indicators from FMP
 */
export interface FinanceIndicators {
  /** VIX volatility index value */
  vix: number | null;
  /** VIX change from previous period */
  vixChange: number | null;
  /** US Dollar Index */
  dxy: number | null;
  /** Yield spread (10Y-2Y) */
  yieldSpread: number | null;
  /** Carry trade indicator */
  carryTrade: number | null;
}

/**
 * Regional escalation heatmap data
 */
export interface RegionHeatmap {
  /** Region name (e.g., "Middle East", "Europe") */
  region: string;
  /** Geopolitical escalation score (0-1) */
  geopolitical: number;
  /** Military escalation score (0-1) */
  military: number;
  /** Economic escalation score (0-1) */
  economic: number;
}

/**
 * Correlation alert between market and news signals
 */
export interface CorrelationAlert {
  /** Alert type */
  type: AlertType;
  /** Human-readable alert message */
  message: string;
  /** ISO timestamp when alert was generated */
  timestamp: string;
  /** Alert confidence (0-1) */
  confidence: number;
  /** Related cluster ID if applicable */
  clusterId?: string;
}

// =============================================================================
// API Response Interfaces
// =============================================================================

/**
 * Escalation summary response from /escalation/summary endpoint
 */
export interface EscalationSummary {
  /** Current market regime status */
  regime: RegimeStatus;
  /** Financial market indicators */
  finance: FinanceIndicators;
  /** Regional escalation heatmap */
  heatmap: RegionHeatmap[];
  /** Active correlation alerts */
  alerts: CorrelationAlert[];
}

/**
 * Signal breakdown by source type
 */
export interface EscalationSignals {
  /** Embedding similarity signals per domain */
  embedding: Record<string, { score: number; level?: number; anchor_id?: string }>;
  /** Content analysis signals per domain */
  content: Record<string, { score: number }>;
  /** Keyword heuristic signals per domain */
  keywords: Record<string, { score: number }>;
}

/**
 * Detailed escalation data for a single cluster
 * Response from /escalation/clusters/{id} endpoint
 */
export interface ClusterEscalation {
  /** Cluster UUID */
  clusterId: string;
  /** Geopolitical escalation score (0-1) */
  geopolitical: number;
  /** Military escalation score (0-1) */
  military: number;
  /** Economic escalation score (0-1) */
  economic: number;
  /** Combined weighted score (0-1) */
  combined: number;
  /** Escalation level (1-5) */
  level: EscalationLevel;
  /** Signal breakdown by source */
  signals: EscalationSignals;
  /** FMP market correlation data */
  fmpCorrelation: {
    alertType: AlertType | null;
    message: string | null;
    regimeAtTime: string | null;
    confidence: number | null;
  } | null;
  /** ISO timestamp when escalation was calculated */
  calculatedAt: string | null;
}

// =============================================================================
// UI Helper Constants
// =============================================================================

/**
 * Escalation level metadata for UI rendering
 */
export const ESCALATION_LEVELS: Record<
  EscalationLevel,
  { label: string; color: string; bgColor: string; description: string }
> = {
  1: {
    label: 'Routine',
    color: 'text-green-600',
    bgColor: 'bg-green-100',
    description: 'Normal activity levels',
  },
  2: {
    label: 'Elevated',
    color: 'text-yellow-600',
    bgColor: 'bg-yellow-100',
    description: 'Increased monitoring recommended',
  },
  3: {
    label: 'Significant',
    color: 'text-orange-600',
    bgColor: 'bg-orange-100',
    description: 'Notable developments requiring attention',
  },
  4: {
    label: 'High',
    color: 'text-red-600',
    bgColor: 'bg-red-100',
    description: 'Critical situation developing',
  },
  5: {
    label: 'Critical',
    color: 'text-gray-100',
    bgColor: 'bg-gray-800',
    description: 'Maximum alert - immediate attention required',
  },
};

/**
 * Regime type colors for UI rendering
 */
export const REGIME_COLORS: Record<RegimeType, string> = {
  RISK_ON: 'bg-green-500',
  RISK_OFF: 'bg-red-500',
  TRANSITIONAL: 'bg-yellow-500',
  UNKNOWN: 'bg-gray-400',
};

/**
 * Alert type colors for UI rendering
 */
export const ALERT_TYPE_COLORS: Record<AlertType, { bg: string; text: string; border: string }> = {
  CONFIRMATION: {
    bg: 'bg-green-50',
    text: 'text-green-700',
    border: 'border-green-200',
  },
  DIVERGENCE: {
    bg: 'bg-red-50',
    text: 'text-red-700',
    border: 'border-red-200',
  },
  EARLY_WARNING: {
    bg: 'bg-amber-50',
    text: 'text-amber-700',
    border: 'border-amber-200',
  },
};

// =============================================================================
// UI Helper Functions
// =============================================================================

/**
 * Get escalation level metadata
 *
 * @param level - Escalation level (1-5)
 * @returns Level metadata with label, colors, and description
 */
export function getEscalationLevelInfo(level: EscalationLevel) {
  return ESCALATION_LEVELS[level] || ESCALATION_LEVELS[1];
}

/**
 * Convert score (0-1) to escalation level (1-5)
 *
 * @param score - Combined score from 0 to 1
 * @returns Escalation level from 1 to 5
 */
export function scoreToLevel(score: number): EscalationLevel {
  if (score < 0.2) return 1;
  if (score < 0.4) return 2;
  if (score < 0.6) return 3;
  if (score < 0.8) return 4;
  return 5;
}

/**
 * Get regime display information
 *
 * @param type - Regime type
 * @returns Object with label and color
 */
export function getRegimeInfo(type: RegimeType): { label: string; color: string; description: string } {
  const info: Record<RegimeType, { label: string; color: string; description: string }> = {
    RISK_ON: {
      label: 'Risk On',
      color: REGIME_COLORS.RISK_ON,
      description: 'Markets show appetite for risk assets',
    },
    RISK_OFF: {
      label: 'Risk Off',
      color: REGIME_COLORS.RISK_OFF,
      description: 'Flight to safety - defensive positioning',
    },
    TRANSITIONAL: {
      label: 'Transitional',
      color: REGIME_COLORS.TRANSITIONAL,
      description: 'Market regime uncertain - watch for signals',
    },
    UNKNOWN: {
      label: 'Unknown',
      color: REGIME_COLORS.UNKNOWN,
      description: 'No market data available',
    },
  };
  return info[type] || info.UNKNOWN;
}

/**
 * Get alert type display information
 *
 * @param type - Alert type
 * @returns Object with label, colors, and description
 */
export function getAlertTypeInfo(type: AlertType): {
  label: string;
  colors: { bg: string; text: string; border: string };
  description: string;
} {
  const info: Record<
    AlertType,
    { label: string; colors: { bg: string; text: string; border: string }; description: string }
  > = {
    CONFIRMATION: {
      label: 'Confirmation',
      colors: ALERT_TYPE_COLORS.CONFIRMATION,
      description: 'Market regime aligns with news escalation',
    },
    DIVERGENCE: {
      label: 'Divergence',
      colors: ALERT_TYPE_COLORS.DIVERGENCE,
      description: 'Market and news signals conflict - investigate',
    },
    EARLY_WARNING: {
      label: 'Early Warning',
      colors: ALERT_TYPE_COLORS.EARLY_WARNING,
      description: 'Market may be leading news signals',
    },
  };
  return info[type];
}

/**
 * Format escalation score as percentage string
 *
 * @param score - Score from 0 to 1
 * @param decimals - Number of decimal places (default: 1)
 * @returns Formatted percentage string (e.g., "45.2%")
 */
export function formatEscalationScore(score: number, decimals: number = 1): string {
  return `${(score * 100).toFixed(decimals)}%`;
}
